src/ImageData.vala

```diff
@@ -0,0 +1,66 @@
+/*
+* Copyright 2020 elementary, Inc. (https://elementary.io)
+*
+* This program is free software; you can redistribute it and/or
+* modify it under the terms of the GNU General Public
+* License as published by the Free Software Foundation; either
+* version 3 of the License, or (at your option) any later version.
+*
+* This program is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+* General Public License for more details.
+*
+* You should have received a copy of the GNU General Public
+* License along with this program; if not, write to the
+* Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+* Boston, MA 02110-1301 USA
+*
+*/
+
+public class Notifications.ImageData : GLib.Object {
+    public int width { get; construct; }
+    public int height { get; construct; }
+    public int rowstride { get; construct; }
+    public bool has_alpha { get; construct; }
+    public int bits_per_sample { get; construct; }
+    public int n_channels { get; construct; }
+    public void* raw { get; set; }
+
+    public ImageData (int width, int height, int rowstride, bool has_alpha, int bits_per_sample, int n_channels, void* raw) {
+        Object (
+            width: width, 
+            height: height, 
+            rowstride: rowstride, 
+            has_alpha: has_alpha, 
+            bits_per_sample: bits_per_sample, 
+            n_channels: n_channels, 
+            raw: raw
+        );
+    }
+
+    /**
+     * Decode a raw image (iiibiiay) sent through 'hints'
+     */
+    public static ImageData from_variant(Variant img)
+    {
+        // Read the image fields
+        int v_width           = img.get_child_value(0).get_int32();
+        int v_height          = img.get_child_value(1).get_int32();
+        int v_rowstride       = img.get_child_value(2).get_int32();
+        bool v_has_alpha      = img.get_child_value(3).get_boolean();
+        int v_bits_per_sample = img.get_child_value(4).get_int32();
+        int v_n_channels      = img.get_child_value(5).get_int32();
+        void* v_raw = img.get_child_value (6).get_data();
```

Comment:
It sounds like the data returned from `get_data` is only valid for as long as the owning `Variant` is alive, and it goes out of scope at the end of this method. I think you want to find a way to decode this part of the variant as a `uint8` array so that this class can have ownership of the data.

Reply:
That's a good point. I'll find a way around this. Something of note, GLib doesn't allow `uint8[]` as a property, so I was using the int as a pointer here. Do you know if there's a better way to store it on the ImageData object?

Reply:
I decided to simplify and just create our own Pixbuf from the image-data. This should prevent any weird memory timing with image data disappearing as we go to show it.